### PR TITLE
DP-34883: Modify form content type to accomodate other form types

### PR DESCRIPTION
**Pull request by arthurbaghdas**
**Description:**
Explain the technical implementation of the work done.

**Jira:** (Skip unless you are MA staff)
DP-****

**To Test:**
- [ ] Add steps to test this feature

---

[Peer Review Checklist](https://github.com/massgov/openmass/blob/develop/docs/peer_review_checklist.md)

### CODE REVIEW

**Comment by tom-fleming** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Add a changelog YAML file to this PR</td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against a346111800c0e3282424d7b363d2322970100ad3
</p>

File: docroot/modules/custom/mass_content/mass_content.deploy.php [hunk 1]
```diff
@@ -257,3 +257,76 @@ function mass_content_deploy_card_label_migration(&$sandbox) {
...
+      }
+    }
+    catch (\Exception $e) {
+      // Catch any exception and continue processing.
+      \Drupal::logger('mass_validation')->error('Failed to update node @nid: @message', [
+        '@nid' => $node->id(),
+        '@message' => $e->getMessage(),
+      ]);
+      // Re-enable entity hierarchy writes in case of an error.
+      \Drupal::state()->set('entity_hierarchy_disable_writes', FALSE);
```
Comment (weitzman) [inline docroot/modules/custom/mass_content/mass_content.deploy.php]: Is the intention to break out of the loop or keep going after an exception. If we are to keep going, then I dont think it makes sense to re-enable the writes

File: docroot/themes/custom/mass_theme/mass_theme.theme [hunk 2]
```diff
@@ -611,25 +611,35 @@ function mass_theme_preprocess_node_form_page(&$variables) {
...
+                '%message' => $e->getMessage(),
+              ]);
+          }
         }
       }
-      catch (\Exception $e) {
-        \Drupal::logger('mass_fields')->warning('Parsing of formstack URL failed: %message', [
-          '%message' => $e->getMessage(),
-        ]);
+      elseif ($node->field_form_platform->value == 'gravity_forms') {
```
Comment (weitzman) [inline docroot/themes/custom/mass_theme/mass_theme.theme]: Please add. method to the bundle class(es) that have this field and get the value using that method. That yields better code and centralizes our access of field properties.

File: docroot/modules/custom/mass_content/mass_content.deploy.php (hunk 1 above)
Comment (arthurbaghdas) [reply docroot/modules/custom/mass_content/mass_content.deploy.php]: Makes sense just removing it, updated the code, thank you

File: docroot/themes/custom/mass_theme/mass_theme.theme (hunk 2 above)
Comment (arthurbaghdas) [reply docroot/themes/custom/mass_theme/mass_theme.theme]: Added and replaced with that method.

File: docroot/themes/custom/mass_theme/mass_theme.theme [hunk 3]
```diff
@@ -611,25 +611,35 @@ function mass_theme_preprocess_node_form_page(&$variables) {
...
-        $validate_script_url = parse_url($script_url, PHP_URL_HOST);
-        $validate_noscript_url = parse_url($noscript_url, PHP_URL_HOST);
-
-        if (str_ends_with($validate_script_url, "formstack.com")) {
-          $variables['form_url'] = Url::fromUri($script_url);
-        }
-        if (str_ends_with($validate_noscript_url, "formstack.com")) {
-          $variables['form_url_noscript'] = $noscript_url;
+    if (!$node->getPlatform()->isEmpty()) {
+      $variables['formPlatform'] = $node->field_form_platform->value;
```
Comment (weitzman) [inline docroot/themes/custom/mass_theme/mass_theme.theme]: IMo these calls are best done at `$node->getPlatform()->getString()`, as you did once below.

File: docroot/themes/custom/mass_theme/mass_theme.theme [hunk 4]
```diff
@@ -611,25 +611,35 @@ function mass_theme_preprocess_node_form_page(&$variables) {
...
+      if ($node->field_form_platform->value == 'formstack') {
+        if (!empty($node->field_form_embed->value)) {
```
Comment (weitzman) [inline docroot/themes/custom/mass_theme/mass_theme.theme]: Ideally, make a bundle class method for this field and use isEmpty()

File: docroot/themes/custom/mass_theme/mass_theme.theme [hunk 5]
```diff
@@ -611,25 +611,35 @@ function mass_theme_preprocess_node_form_page(&$variables) {
...
+                '%message' => $e->getMessage(),
+              ]);
+          }
         }
       }
-      catch (\Exception $e) {
-        \Drupal::logger('mass_fields')->warning('Parsing of formstack URL failed: %message', [
-          '%message' => $e->getMessage(),
-        ]);
+      elseif ($node->getPlatform()->getString() == 'gravity_forms') {
```
Comment (weitzman) [inline docroot/themes/custom/mass_theme/mass_theme.theme]: Lets make string constants for `gravity_forms` and `formstack` so we avoid typos and can search for usages.

**Comment by arthurbaghdas** [comment]
Can you review this agaiin @weitzman ?

File: docroot/modules/custom/mass_content/src/Entity/Bundle/node/FormPageBundle.php [hunk 6]
```diff
@@ -2,9 +2,65 @@
...
+  public function isFormStack(): bool {
+    $platform = $this->getPlatform();
+    return !empty($platform->getString()) && $platform->getString() === self::PLATFORM_FORMSTACK;
+  }
+
+  /**
+   * Check if platform is Gravity Forms.
+   *
+   * @return bool
+   *   TRUE if the platform is Gravity Forms and field is set, FALSE otherwise.
```
Comment (weitzman) [inline docroot/modules/custom/mass_content/src/Entity/Bundle/node/FormPageBundle.php:38]: FYI, its OK IMO to omit obvious PHPDoc. I think our PHPCS is configured to allow this.

File: docroot/themes/custom/mass_theme/mass_theme.theme [hunk 7]
```diff
@@ -611,25 +611,39 @@ function mass_theme_preprocess_node_form_page(&$variables) {
...
-        $validate_script_url = parse_url($script_url, PHP_URL_HOST);
-        $validate_noscript_url = parse_url($noscript_url, PHP_URL_HOST);
-
-        if (str_ends_with($validate_script_url, "formstack.com")) {
-          $variables['form_url'] = Url::fromUri($script_url);
-        }
-        if (str_ends_with($validate_noscript_url, "formstack.com")) {
-          $variables['form_url_noscript'] = $noscript_url;
+    if (!$node->getPlatform()->isEmpty()) {
+      $variables['formPlatform'] = $node->getPlatform()->getString();
```
Comment (weitzman) [inline docroot/themes/custom/mass_theme/mass_theme.theme]: Now that you have a getter on the bundle class, you dont need to create variables during preprocess anymore. In the twig file, call `{{ node.getPlatform() }}`. Search for 'twig' at https://www.lullabot.com/articles/drupals-bundle-classes-empower-better-code for more information.

Comment (arthurbaghdas) [reply docroot/themes/custom/mass_theme/mass_theme.theme]: Done, thanks

File: docroot/themes/custom/mass_theme/templates/content/node--form-page.html.twig [hunk 8]
```diff
@@ -69,21 +69,31 @@
...
-          var root = document.querySelector('.form_embed.feedback');
-          root.addEventListener('click', function(event) {
-            //find target element, if submit button, scroll to top of form.
-            if(event.target.tagName === 'INPUT' && event.target.type === 'submit') {
-              root.scrollIntoView();
-            }
-          });
-        })();
-      </script>
+      {% if node.getPlatform().getString() == 'formstack' %}
```
Comment (weitzman) [inline docroot/themes/custom/mass_theme/templates/content/node--form-page.html.twig]: Looks like you can use the new class constants in Twig (one more replacement is below this line) - https://stackoverflow.com/a/12990418/265501

File: docroot/modules/custom/mass_content/src/Entity/Bundle/node/FormPageBundle.php [hunk 9]
```diff
@@ -2,9 +2,65 @@
...
+    return !empty($platform->getString()) && $platform->getString() === self::PLATFORM_GRAVITY_FORMS;
+  }
+
+  /**
+   * Get form embed field.
+   *
+   * @return string|null
+   *   The value of the form embed field, or NULL if it's empty.
+   */
+  public function getFormEmbed(): ?string {
```
Comment (weitzman) [inline docroot/modules/custom/mass_content/src/Entity/Bundle/node/FormPageBundle.php]: Most other getters in bundle class return `FieldItemListInterface` and expect the caller to call getString(). I think it makes sense to do that here.

Comment (arthurbaghdas) [reply docroot/modules/custom/mass_content/src/Entity/Bundle/node/FormPageBundle.php]: ok

**Review by weitzman — APPROVED** [review]
LGTM

File: docroot/modules/custom/mass_validation/mass_validation.module [hunk 10]
```diff
@@ -55,25 +55,37 @@ function mass_validation_entity_type_alter(array &$entity_types) {
...
+          'blog.mass.gov',
+          'www.mass.gov',
+        ],
+      ]);
+    }
+    if (array_key_exists('field_form_url', $fields) && !empty($fields['field_form_url'])) {
+      $field_form_url = &$fields['field_form_url'];
+      $field_form_url->addConstraint('GravityForm', [
+        'allowedValues' => [
+          'massforms-gravity.flywheelsites.com',
```
Comment (joeg8612) [inline docroot/modules/custom/mass_validation/mass_validation.module]: We need to change the domains.
Allow:
- live-mass-forms.pantheonsite.io
- forms.mass.gov

Remove:
- massforms-gravity.flywheelsites.com

**Comment by tom-fleming** [comment]
LGTM

**Pull request merged**